lib/segment/src/vector_storage/async_io.rs

```diff
@@ -63,24 +63,36 @@ impl UringReader {
         Ok(Self {
             file,
             buffers,
-            io_uring,
+            io_uring: Some(io_uring),
             raw_size,
             header_size,
         })
     }
 
+    pub fn drop_io_uring(&mut self) {
+        self.io_uring = None;
+    }
+
+    fn ensure_io_uring(&mut self) -> OperationResult<()> {
+        if self.io_uring.is_none() {
+            self.io_uring = Some(IoUring::new(DISK_PARALLELISM as _)?);
+        }
+        Ok(())
+    }
+
     fn submit_and_read(
         &mut self,
         mut callback: impl FnMut(usize, PointOffsetType, &[VectorElementType]),
         unused_buffer_ids: &mut Vec<usize>,
     ) -> OperationResult<()> {
         let buffers_count = self.buffers.buffers.len();
         let used_buffers_count = buffers_count - unused_buffer_ids.len();
+        let io_uring = self.io_uring.as_mut().unwrap();
```

Comment:
this unwrap is ugly, yes, but I didn't find a way to do it better because it doesn't allow to borrow self mutable twice

Reply:
> Or do you want to unwrap here explicitly so we error in case uring was dropped due to an error?

As far as I can see - this. The code is structured in a way, so that `submit_and_read` is called when we already have `io_uring`, so `ensure_io_uring` is both unnecessary and undesirable.

Reply:
Maybe we can add a comment or use `expect()` to elaborate on this. I imagine that to be useful if we ever hit this.

Something like:

```suggestion
        let io_uring = self.io_uring.as_mut().expect("unrecoverable io_uring state");
```

Reply:
Reworked this bit.